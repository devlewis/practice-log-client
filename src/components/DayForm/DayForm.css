.italics {
  font-style: italic;
  font-size: 1em;
}

.selects {
  margin: 1em 0 1.5em;
}

.required_box {
  margin-bottom: 1em;
}

.required_box h2 {
  margin: .5em 0;
}

.optional_box {
  border: solid black 2px;
  margin-bottom: 10px;
  padding: 2px;
  margin-right: 0;
}

.error {
  color: red;
  font-size: 10px;
}

select {
  border: 1px solid black;
  margin-left: 5px;
}

.DayForm__form input {
  margin-bottom: 10px;
}

.optional_box p {
  margin-top: 0;
}

#actual_hours {
  width: 50px;
  margin-left: 5px;
  border: 1px solid gray;
}

#technique,
#repertoire {
  margin-top: 3px;
  margin-left: 3px;
  width: 96%;
  border: 1px solid gray;
}

.DayForm__form {
  width: 250px;
  margin: auto;
  background-color: aliceblue;
  padding: 20px;
  margin-top: 0px;
}

.DayForm__form h1 {
  margin: 0;
}

.DayForm__buttons {
  display: flex;
  justify-content: space-between;
}

.DayForm__buttons button:first-child {
  background-color: dimgray;
}

@media screen and (min-width: 321px) {
  .DayForm__form {
    margin-top: 25px;
    width: 300px;
  }
}

@media screen and (min-width: 415px) {
  select {
    font-size: 0.9em;
  }
}

@media screen and (min-width: 768px) {
  .DayForm__form {
    width: 500px;
  }
}
