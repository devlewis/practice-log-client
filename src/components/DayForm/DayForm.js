import React, { Component } from "react";
//import PropTypes from "prop-types";
import "./DayForm.css";
import Context from "../../Context";
import ValidationError from "../Utils/ValidationError";

class DayForm extends Component {
  static propTypes = {};

  static contextType = Context;

  state = {
    error: null,
    id: "",
    day_num: "",
    day_date: "",
    completed: "",
    technique: "",
    repertoire: "",
    actual_hours: { value: "", touched: false },
    touched: "",
    goal_id: "",
    user_id: "",
    goal_hours: "",
  };

  componentDidMount() {
    const dayNum = this.props.location.pathname.split("/")[2];
    const day = this.context.days[dayNum - 1];

    this.setState({
      id: day.id,
      day_num: day.day_num,
      day_date: day.date,
      completed: day.completed,
      technique: day.technique,
      repertoire: day.repertoire,
      actual_hours: { value: day.actual_hours },
      touched: day.touched,
      goal_id: day.goal_id,
      user_id: day.user_id,
      goal_hours: day.goal_hours,
    });
  }

  handleSubmit = (e) => {
    e.preventDefault();
    const {
      id,
      day_num,
      day_date,
      completed,
      technique,
      repertoire,
      actual_hours,
      goal_id,
      user_id,
      goal_hours,
    } = this.state;

    const updatedDay = {
      id,
      day_num,
      day_date,
      completed,
      technique,
      repertoire,
      actual_hours: parseFloat(actual_hours.value.toFixed(2)),
      touched: true,
      goal_id,
      user_id,
      goal_hours,
    };

    const dayId = id;

    this.context.onHandleDaySubmit(updatedDay, dayId);

    this.props.history.goBack();
  };

  validateHours() {
    const hours = this.state.actual_hours.value;
    if (isNaN(hours) || hours < 0 || hours > 24) {
      return "please enter a number of hours between 0 and 24";
    }
  }

  handleChangeCompleted = (e) => {
    this.setState({
      completed: e.target.value,
      actual_hours: { value: this.state.goal_hours },
    });
  };

  handleChangeHours = (e) => {
    this.setState({
      actual_hours: { value: parseFloat(e.target.value), touched: true },
    });
  };

  handleChangeTechnique = (e) => {
    this.setState({ technique: e.target.value });
  };

  handleChangeRepertoire = (e) => {
    this.setState({ repertoire: e.target.value });
  };

  render() {
    const hoursError = this.validateHours();
    const dayNum = this.props.location.pathname.split("/")[2];
    const { error } = this.state;
    const { id, completed, technique, repertoire, actual_hours } = this.state;

    return (
      <form className="DayForm__form" onSubmit={this.handleSubmit}>
        <div className="DayForm__error" role="alert">
          {error && <p>{error.message}</p>}
        </div>
        {id && <input type="hidden" name="id" value={id} />}
        <div className="required_box">
          <h1>Practice Day #{dayNum}</h1>
          <h2>{this.state.day_date}</h2>
          <h3>Goal Number of Hours:{this.state.goal_hours}</h3>
          <div className="error" role="alert">
            {error && <p>{error.message}</p>}
          </div>
          <label htmlFor="completed">Did you practice today?</label>
          <select
            onChange={this.handleChangeCompleted}
            name="completed"
            id="completed"
            required
            value={completed}
          >
            <option value="false">No. :-(</option>
            <option value="true">Yes!!!</option>
          </select>
        </div>
        {this.state.completed === "true" && (
          <section>
            <div>
              <label htmlFor="actual_hours">How Many Hours?</label>
              <input
                type="number"
                name="actual_hours"
                id="actual_hours"
                value={actual_hours.value}
                onChange={this.handleChangeHours}
              />
              {this.state.actual_hours.touched && (
                <ValidationError message={hoursError} />
              )}
            </div>
            <div className="optional_box">
              <p className="italics">Optional:</p>

              <div>
                <label htmlFor="technique">Technique:</label>
                <input
                  type="text"
                  name="technique"
                  id="technique"
                  placeholder="ex. Hanon No. 1"
                  value={technique}
                  onChange={this.handleChangeTechnique}
                />
              </div>
              <div>
                <label htmlFor="repertoire">Repertoire:</label>
                <input
                  type="text"
                  name="repertoire"
                  id="repertoire"
                  value={repertoire}
                  onChange={this.handleChangeRepertoire}
                />
              </div>
            </div>
          </section>
        )}
        <div className="DayForm__buttons">
          <button type="button" onClick={(e) => this.props.history.goBack()}>
            Cancel
          </button>{" "}
          <button type="submit">Save</button>
        </div>
      </form>
    );
  }
}

export default DayForm;
