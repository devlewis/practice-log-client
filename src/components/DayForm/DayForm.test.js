import React from "react";
import { shallow, mount } from "enzyme";
import toJson from "enzyme-to-json";
import DayForm from "./DayForm";
import Context from "../../Context";

let value = {
  days: [
    {
      id: 1,
      day_num: 1,
      date: "March 22",
      completed: "false",
      technique: "",
      repertoire: "",
      actual_hours: 0,
      touched: false,
      goal_id: 1,
      user_id: 1,
      goal_hours: 3,
    },
    {
      id: 2,
      day_num: 2,
      date: "March 23",
      completed: "true",
      technique: "Hanon",
      repertoire: "",
      actual_hours: 2,
      touched: true,
      goal_id: 1,
      user_id: 1,
      goal_hours: 3,
    },
  ],
};

describe(`DayForm component`, () => {
  const props = {
    location: { pathname: "/dayform/3" },
    history: { goBack: () => {} },
    className: "test-class-name",
    children: <p>test children</p>,
    "data-other": "test-other-prop",
  };

  it("renders a form.DayForm by default", () => {
    const wrapper = shallow(<DayForm />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("renders the DayForm given props", () => {
    const wrapper = mount(
      <Context.Provider value={value}>
        <DayForm {...props} />
      </Context.Provider>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
