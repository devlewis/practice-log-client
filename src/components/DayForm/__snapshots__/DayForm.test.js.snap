// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DayForm component renders a form.DayForm by default 1`] = `
Array [
  <DayForm
    className="test-class-name"
    data-other="test-other-prop"
    history={
      Object {
        "goBack": [Function],
      }
    }
    location={
      Object {
        "pathname": "/dayform/1",
      }
    }
  >
    <form
      className="DayForm__form"
      onSubmit={[Function]}
    >
      <div
        className="DayForm__error"
        role="alert"
      />
      <input
        name="id"
        type="hidden"
        value={1}
      />
      <div
        className="required_box"
      >
        <h1>
          Practice Day #
          1
        </h1>
        <h2>
          March 22
        </h2>
        <h3>
          Goal Number of Hours:
          3
        </h3>
        <div
          className="error"
          role="alert"
        />
        <label
          htmlFor="completed"
        >
          Did you practice today?
        </label>
        <select
          id="completed"
          name="completed"
          onChange={[Function]}
          required={true}
          value="false"
        >
          <option
            value="false"
          >
            No. :-(
          </option>
          <option
            value="true"
          >
            Yes!!!
          </option>
        </select>
      </div>
      <div
        className="DayForm__buttons"
      >
        <button
          onClick={[Function]}
          type="button"
        >
          Cancel
        </button>
         
        <button
          type="submit"
        >
          Save
        </button>
      </div>
    </form>
  </DayForm>,
  ")",
]
`;

exports[`DayForm component renders the DayForm given props 1`] = `
<DayForm
  className="test-class-name"
  data-other="test-other-prop"
  history={
    Object {
      "goBack": [Function],
    }
  }
  location={
    Object {
      "pathname": "/dayform/1",
    }
  }
>
  <form
    className="DayForm__form"
    onSubmit={[Function]}
  >
    <div
      className="DayForm__error"
      role="alert"
    />
    <input
      name="id"
      type="hidden"
      value={1}
    />
    <div
      className="required_box"
    >
      <h1>
        Practice Day #
        1
      </h1>
      <h2>
        March 22
      </h2>
      <h3>
        Goal Number of Hours:
        3
      </h3>
      <div
        className="error"
        role="alert"
      />
      <label
        htmlFor="completed"
      >
        Did you practice today?
      </label>
      <select
        id="completed"
        name="completed"
        onChange={[Function]}
        required={true}
        value="false"
      >
        <option
          value="false"
        >
          No. :-(
        </option>
        <option
          value="true"
        >
          Yes!!!
        </option>
      </select>
    </div>
    <div
      className="DayForm__buttons"
    >
      <button
        onClick={[Function]}
        type="button"
      >
        Cancel
      </button>
       
      <button
        type="submit"
      >
        Save
      </button>
    </div>
  </form>
</DayForm>
`;
